src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceBeforeCaseDefaultColonCheck.java

```diff
@@ -0,0 +1,194 @@
+ *             break;
+ *         case 3, 4
+ *                  : break; // violation, ':' should not be in separate line
+ *         case 4,
+ *               5: break; // ok
+ *         default
+ *               : // violation, ':' should not be in separate line
+ *             break;
+ *     }
+ *   }
+ * }
+ * </pre>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code ws.colonInSeparateLine}
+ * </li>
+ * <li>
+ * {@code ws.preceded}
+ * </li>
+ * </ul>
+ *
+ * @since 8.42
+ */
+@StatelessCheck
+public class NoWhitespaceBeforeCaseDefaultColonCheck
+    extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "ws.preceded";
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY_ON_SEPARATE = "ws.colonInSeparateLine";
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {TokenTypes.COLON};
+    }
+
+    @Override
+    public boolean isCommentNodesRequired() {
+        return true;
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final DetailAST parent = ast.getParent();
+        final boolean isColonOfCase = parent.getType() == TokenTypes.LITERAL_CASE;
+        final boolean isColonOfDefault = parent.getType() == TokenTypes.LITERAL_DEFAULT;
+        final boolean violation = isColonOfCase && checkWhitespaceBeforeColonOfCase(ast)
+                || isColonOfDefault && checkWhitespaceBeforeColonOfDefault(ast);
+        if (violation) {
+            log(ast, MSG_KEY, ast.getText());
+        }
+    }
+
+    /**
+     * Checks if there is a whitespace before the colon of a switch case.
+     *
+     * @param colonAst DetailAST to check.
+     * @return whether there is a whitespace before colonAst or not.
+     */
+    private boolean checkWhitespaceBeforeColonOfCase(DetailAST colonAst) {
+        final DetailAST previousSibling = colonAst.getPreviousSibling();
+        final boolean areOnSameLine = TokenUtil.areOnSameLine(previousSibling, colonAst);
+        boolean horizontalWhitespaceDetected = false;
+        if (areOnSameLine) {
+            int offset = 0;
+            if (previousSibling.getType() == TokenTypes.BLOCK_COMMENT_BEGIN) {
+                offset = 1;
+            }
+            horizontalWhitespaceDetected =
+                    colonAst.getColumnNo() != getLastColumnNumberOf(previousSibling) + offset;
+        }
+        else {
+            log(colonAst, MSG_KEY_ON_SEPARATE);
+        }
+        return horizontalWhitespaceDetected;
```

Comment:
I think that the duplication of code for checking if the tokens are on different lines is unnecessary; can we extract this and make a new method just for checking across lines? So we have  `checkWhitespaceBeforeColonOfDefault()`, `checkWhitespaceBeforeColonOfCase()`, then maybe `checkIfColonIsSameLine()` or something similar, all of which are called from `visitToken()`?.

Reply:
This definitely makes the code clean, but we cannot use here since we comparing different tokens with `colonAst` for same line. I mean we can do but then there would be some more extra conditions to check like if something is not null and all.

Reply:
My thoughts were something like this in `visitToken()`:

```java
        if (!isColonOnSameLine) {
            log();
        } 
        else if (isWhitespacePrecedingColonOfDefault() 
              || isWhitespacePrecedingColonOfCase()) {
           log();
       }
```

I think it is a bit confusing to have a boolean method that also logs violations.  We should at least specify in the javadocs for that these methods also log violations for tokens on different lines, but I think it would be best to keep this code separate.  @romani @strkkk @pbludov please share your thoughts on this and below suggestions.

Reply:
@romani @strkkk @pbludov ping

Reply:
I do not like current way. It seems to me that we can log violation in method and then log it one more time after we returned something from method.
It is better to log it in one place as @nmancus1 suggested

Reply:
Done.

Reply:
@nmancus1 @strkkk Semaphore is failing on `ci/no-exception-test.sh guava-with-google-checks`, Error log says:
```
Error rendering Maven report: Failed during checkstyle configuration: cannot initialize module TreeWalker - cannot initialize
module NoWhitespaceBeforeCaseDefaultColon - Unable to instantiate 'NoWhitespaceBeforeCaseDefaultColon'
class, it is also not possible to instantiate it as 
com.github.sevntu.checkstyle.checks.annotation.NoWhitespaceBeforeCaseDefaultColon,
...
```
Why should sevntu check config being used? Is it introduced in #9278?

Reply:
> Why should sevntu check config being used? 

This is just an error message from the  mvn-checkstyle-plugin.

> Is it introduced in #9278?

The problem is that while diff.groovy doesn't use the branch specified in single mode, it still checks it out. So, this CI job ran on master instead of your PR branch. PR to fix this is at https://github.com/checkstyle/checkstyle/pull/9868. Once that is merged, rebase this PR.

Reply:
Ok, thanks.

Reply:
Thanks for catching this. The reason that we are trying to stop using launch.groovy is that it  is deprecated and will be removed  soon.

Reply:
@AkMo3 , please rebase

Reply:
Resolved.